kml/minzoom_quadtree.hpp

```diff
@@ -0,0 +1,177 @@
+    {
+      auto const [x, y] = elem.m_point;
+      auto const gridX = static_cast<uint32_t>(std::trunc((x - minX) * gridScale));
+      auto const gridY = static_cast<uint32_t>(std::trunc((y - minY) * gridScale));
+      // Sacrifice one of 32 possible levels for simplicity of traversing.
+      elem.m_zCurveCoord = bits::BitwiseMerge(gridX, gridY) >> 2;
+    }
+
+    std::sort(m_quadtree.begin(), m_quadtree.end());
+
+    auto const traverse = [&](auto const & traverse, auto const treeBeg, auto const treeEnd,
+                              uint64_t treeLevelMask, int zoom) -> Element *
+    {
+      if (treeBeg == treeEnd)
+        return nullptr;
+      if (std::next(treeBeg) == treeEnd)
+        return &*treeBeg;
+      if (++zoom >= maxZoom)
+      {
+        auto const rankLess = [&](auto const & lhs, auto const & rhs) -> bool
+        {
+          return m_less(lhs.m_value, rhs.m_value);
+        };
+        auto const result = std::max_element(treeBeg, treeEnd, rankLess);
+        auto const setMaxZoom = [&](auto & elem) { setMinZoom(elem.m_value, maxZoom); };
+        std::for_each(treeBeg, result, setMaxZoom);
+        std::for_each(std::next(result), treeEnd, setMaxZoom);
+        return &*result;
+      }
+
+      Element * topRanked = nullptr;
+      treeLevelMask >>= 2;
+      CHECK_NOT_EQUAL(treeLevelMask, 0, ());
+      uint64_t const quadIncrement = treeLevelMask & (treeLevelMask >> 1);
+      uint64_t quadIndex = 0;
+      auto const quadMaskEqual = [&](auto const & elem) -> bool
+      {
+        return (elem.m_zCurveCoord & treeLevelMask) == quadIndex;
+      };
+      auto quadBeg = treeBeg;
+      for (int quadrant = 0; quadrant < 4; ++quadrant)
+      {
+        ASSERT(std::is_partitioned(quadBeg, treeEnd, quadMaskEqual), ());
+        auto const quadEnd = std::partition_point(quadBeg, treeEnd, quadMaskEqual);
+        if (auto const elem = traverse(traverse, quadBeg, quadEnd, treeLevelMask, zoom))
+        {
+          if (topRanked == nullptr)
+            topRanked = elem;
+          else if (m_less(topRanked->m_value, elem->m_value))
+            setMinZoom(std::exchange(topRanked, elem)->m_value, std::max(zoom, 1));
+          else
+            setMinZoom(elem->m_value, std::max(zoom, 1));
+        }
+        quadBeg = quadEnd;
+        quadIndex += quadIncrement;
+      }
+      ASSERT(quadBeg == treeEnd, ());
+      return topRanked;
+    };
+    auto const rootMask = uint64_t{0b11} << (std::numeric_limits<uint64_t>::digits - 2);
+    if (auto const elem = traverse(traverse, m_quadtree.begin(), m_quadtree.end(), rootMask, baseZoom))
+      setMinZoom(elem->m_value, 1);
+    else
+      CHECK(false, (m_quadtree.size()));
+  }
+
+private:
+  struct Element
+  {
+    Element(m2::PointD const & point, Value const & value)
+      : m_point{point}
+      , m_value{value}
+    {}
+
+    Element(m2::PointD const & point, Value && value)
+      : m_point{point}
+      , m_value{std::move(value)}
```

Comment:
You've already asked me to not to use `ValueType` for `typename` of templated `Add` function. I solved this problem by splitting construction of `Element` from forwarding reference to member of aggregate into the pair of `Element`'s constructors. Previously it was an aggregate. Now it is not an aggregate and have copy- and move-constructor and `emplace_back` can be used instead of `push_back(Element{....});`. Name `ValueType` is not used. Both points are satisfied.
Should I rollback these changes?

Reply:
Ah, ok. Leave this as it is now.